File: web/backend/NodeJS-services/database/models/AttendanceLog.js
const mongoose = require('mongoose');

/**
 * AttendanceLog Schema
 * Represents attendance records for scheduled classes
 */
const AttendanceLogSchema = new mongoose.Schema({
  attendanceId: {
    type: Number,
    required: true,
    unique: true,
    default: () => Math.floor(Date.now() / 1000)
  },
  userId: {
    type: String,
    required: true,
    ref: 'UserAccount'
  },
  rfidId: {
    type: String,
    index: true
  },
  scheduleId: {
    type: Number,
    ref: 'ClassSchedule'
  },
  checkInFace: {
    type: String // Path to face image
  },
  checkIn: {
    type: Date
  },
  status: {
    type: String,
    enum: ['Present', 'Late', 'Absent', 'Not Now'],
    default: 'Not Now'
  },
  deviceId: {
    type: Number,
    ref: 'Device'
  },
  classId: {
    type: Number,
    ref: 'Class'
  },
  className: String,
  subjectId: {
    type: Number,
    ref: 'Subject'
  },
  subjectName: String,
  teacherId: {
    type: Number,
    ref: 'Teacher'
  },
  teacherName: String,
  classroomId: {
    type: Number,
    ref: 'Classroom'
  },
  classroomName: String,
  semesterNumber: {
    type: Number
  },
  userRole: {
    type: String,
    enum: ['student', 'teacher', 'admin']
  },
  note: String,
  checkedBy: {
    type: String,
    enum: ['teacher', 'jetson', 'rfid', 'manual'],
    default: 'manual'
  }
}, {
  timestamps: true,
  toJSON: { virtuals: true },
  toObject: { virtuals: true },
  collection: 'AttendanceLog'
});

// Index for efficient queries
AttendanceLogSchema.index({ userId: 1, timestamp: -1 });
AttendanceLogSchema.index({ rfidId: 1, timestamp: -1 });
AttendanceLogSchema.index({ classId: 1, scheduleId: 1, timestamp: -1 });

// Virtual for getting user information
AttendanceLogSchema.virtual('user', {
  ref: 'UserAccount',
  localField: 'userId',
  foreignField: 'userId',
  justOne: true
});

// Index for faster queries
AttendanceLogSchema.index({ userId: 1, scheduleId: 1 });
AttendanceLogSchema.index({ classId: 1, date: 1 });
AttendanceLogSchema.index({ subjectId: 1 });
AttendanceLogSchema.index({ semesterNumber: 1 });
AttendanceLogSchema.index({ deviceId: 1 });

// Virtual for getting rfid information
AttendanceLogSchema.virtual('rfid', {
  ref: 'RFID',
  localField: 'rfidId',
  foreignField: 'RFID_ID',
  justOne: true
});

// Virtual for getting class information
AttendanceLogSchema.virtual('class', {
  ref: 'Class',
  localField: 'classId',
  foreignField: 'classId',
  justOne: true
});

// Virtual for getting schedule information
AttendanceLogSchema.virtual('schedule', {
  ref: 'ClassSchedule',
  localField: 'scheduleId',
  foreignField: 'scheduleId',
  justOne: true
});

// Virtual for getting device information
AttendanceLogSchema.virtual('device', {
  ref: 'Device',
  localField: 'deviceId',
  foreignField: 'deviceId',
  justOne: true
});

// Thêm virtual để populate thông tin student
AttendanceLogSchema.virtual('student', {
  ref: 'Student',
  localField: 'userId',
  foreignField: 'userId',
  justOne: true
});

module.exports = mongoose.model('AttendanceLog', AttendanceLogSchema); 

File: web/backend/Python-services/src/models/AttendanceLog.py
"""
AttendanceLog model for FAMS
"""
from datetime import datetime
from typing import Optional
from pydantic import BaseModel, Field
from ..constants import COLLECTIONS


class AttendanceLog(BaseModel):
    """AttendanceLog model representing attendance records"""
    attendanceId: int = Field(..., description="Unique identifier for the attendance record")
    scheduleId: int = Field(..., description="Reference to ClassSchedule")
    userId: str = Field(..., description="Reference to UserAccount")
    checkInFace: Optional[bytes] = Field(None, description="Face image used for check-in")
    checkIn: Optional[datetime] = Field(None, description="Time of check-in")
    note: Optional[str] = Field("", description="Additional notes for this attendance")
    status: str = Field(default="Absent", description="Attendance status (Present, Late, Absent, Not Now)")
    semesterNumber: Optional[int] = Field(None, description="Semester number for easier filtering")
    userRole: Optional[str] = Field(None, description="Role of the user (teacher, student)")
    teacherId: Optional[int] = Field(None, description="Teacher ID for the class session")
    teacherName: Optional[str] = Field(None, description="Teacher name for display")
    subjectId: Optional[int] = Field(None, description="Subject ID for the class session")
    subjectName: Optional[str] = Field(None, description="Subject name for display")
    classId: Optional[int] = Field(None, description="Class ID for the session")
    className: Optional[str] = Field(None, description="Class name for display")
    studentId: Optional[int] = Field(None, description="Student ID (if applicable)")
    studentName: Optional[str] = Field(None, description="Student name (if applicable)")
    classroomId: Optional[int] = Field(None, description="Classroom ID where the session takes place")
    classroomName: Optional[str] = Field(None, description="Classroom name for display")
    createdAt: datetime = Field(default_factory=datetime.utcnow)
    updatedAt: datetime = Field(default_factory=datetime.utcnow)
    isActive: bool = Field(default=True)
    collection: str = COLLECTIONS['ATTENDANCE_LOG'] 